File: dirs_links/t_unlink/t_unlink.c
#include <fcntl.h>
#include "../../lib/tlpi_hdr.h"

#define CMD_SIZE 200
#define BUF_SIZE 1024

int main(int ac, char *av[])
{
	// Command to be passed to system(3)
	char shell_cmd[CMD_SIZE];
	// random bytes to write to file
	char buf[BUF_SIZE];

	if (ac < 2 || strcmp(av[1], "--help") == 0) {
		usageErr("%s temp-file [num-1kb-blocks]\n", av[0]);
	}

	int num_blks = (ac > 2) ? getInt(av[2], GN_GT_0, "num-1kb-blocks") :
				  100000;

	int fd = open(av[1], O_WRONLY | O_CREAT | O_EXCL, S_IROTH | S_IWUSR);
	if (fd == -1) {
		errExit("open");
	}

	if (unlink(av[1]) == -1) {
		errExit("unlink");
	}

	for (int j = 0; j < num_blks; j += 1) {
		if (write(fd, buf, BUF_SIZE) != BUF_SIZE) {
			fatal("partial/failed write");
		}
	}

	snprintf(shell_cmd, CMD_SIZE, "df -k `dirname %s`", av[1]);
	system(shell_cmd);

	// file is destroyed now
	if (close(fd) == -1) {
		errExit("close");
	}

	printf("********** Closed file descriptor\n");

	system(shell_cmd);
	exit(EXIT_SUCCESS);
}

File: dirs_links/t_unlink/t_unlink.rs
use std::{
    env::args,
    ffi::OsString,
    fs::{remove_file, File},
    io::Write,
    os::unix::ffi::OsStringExt,
    process::{exit, Command},
};

const BUF_SIZE: usize = 1024;

fn main() {
    let mut av = args().skip(1);
    let temp_file = av.next();
    let num_blks = av.next();
    let buf = [0; BUF_SIZE];

    let temp_file = match temp_file {
        Some(file_name) if file_name != "--help" => file_name,
        _ => {
            eprintln!("t_unlink temp-file [num-1kb-blocks]");
            exit(1);
        }
    };

    let num_blks = if num_blks.is_none() {
        1000000
    } else {
        num_blks.unwrap().parse().unwrap()
    };

    let mut file = File::create(temp_file.as_str()).expect("can not create file");
    remove_file(temp_file.as_str()).expect("can not delete file");

    for _ in 0..num_blks {
        assert!(file.write(&buf).unwrap().eq(&BUF_SIZE));
    }

    let mut out = Command::new("dirname")
        .arg(temp_file)
        .output()
        .unwrap()
        .stdout;
    out.truncate(out.len() - 1);
    let dir = OsString::from_vec(out);

    Command::new("df")
        .arg("-k")
        .arg(dir.as_os_str())
        .spawn()
        .unwrap()
        .wait()
        .unwrap();

    drop(file);
    println!("********** Closed file descriptor");

    Command::new("df")
        .arg("-k")
        .arg(dir.as_os_str())
        .spawn()
        .unwrap()
        .wait()
        .unwrap();
}
